Resolved IDENTITY-2558

IndunilRathnayake opened the pull request.

johannnallathamby commented:
I think the issue is misunderstood. The issue is that if default authenticator is updated in the IDP UI, it is not updated in the Advanced Configuration of the SP UI, where we add a step, and then can add a federated IDP (the label is wrong, which says "Federated Authenticator") to that step, and then choose an authenticator if the default is not suitable. The default authenticator shown here is what needs to be updated. If its an already saved configuration of the SP, with a different authenticator than the default, then we must not switch to the default at runtime in the authentication framework, which is what it seems you have done IINM.

johannnallathamby commented on components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/IdentityProviderManager.java:
```diff
@@ -1469,4 +1469,11 @@ private boolean validateUpdateOfIdPEntityId(FederatedAuthenticatorConfig[] curre
+                idpName, getTenantIdOfDomain(tenantDomain), tenantDomain);
+        return federatedAuthenticatorConfig;
+    }
+
```
No need to use separate method here to get default authenticator. The client can use getIdPByName and then pick the default from the returned object. Also the method name is not right, it doesn't contain "default" anywhere.

IndunilRathnayake commented:
Issue: When creating a SP with the federation type as "Federated Authentication" in Local & Outbound Authentication Configuration, the selected federated IDP is saved with current default authenticator. If the default authenticator of IDP is changed afterwards, it is not reflected in SP side. So that when do authentication in SP application, it always authenticated through the saved default authenticator of federated IDP.
Fix: Added a listener to change the IDP's default authenticator in SP side, at the moment of changing IDP configurations. 

Other changes: 
1) added a new Exception in idp-mgt module (previously it uses "IdentityApplicationManagementException" defines in application-mgt)
2) added pre and post methods in ApplicationMgtListener and IdentityProviderMgtLister

johannnallathamby commented on components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java:
```diff
@@ -115,7 +115,9 @@ public int createApplication(ServiceProvider serviceProvider, String tenantDomai
             List<ApplicationMgtListener> listeners = ApplicationMgtListenerServiceComponent.getListners();
```
Let's change return type of createApplication to void. The int value returned is the auto-generated table ID which should not be returned, and isn't used anywhere in the client side as well.

johannnallathamby commented on components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java:
```diff
@@ -115,7 +115,9 @@ public int createApplication(ServiceProvider serviceProvider, String tenantDomai
 
             for (ApplicationMgtListener listener : listeners) {
-                listener.createApplication(serviceProvider);
+                if(!listener.doPreCreateApplication(serviceProvider)){
```
Pre and Post listeners must be called outside the try-catch blocks.

johannnallathamby commented on components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/IdentityProviderDAOImpl.java:
```diff
@@ -59,7 +60,7 @@ public String getDefaultAuthenticator(String idpName)
                     .getDefaultAuthenticatorConfig().getName() : null;
-        } catch (IdentityApplicationManagementException e) {
+        } catch (IdentityProviderManagementException e) {
             throw new IdentityApplicationManagementException(e.getMessage(), e);
```
Better not use e.getMessage(). Instead use a contextual message.

johannnallathamby commented on components/idp-mgt/org.wso2.carbon.idp.mgt/pom.xml:
```diff
@@ -52,6 +52,10 @@
         </dependency>
         <dependency>
             <groupId>org.wso2.carbon.identity</groupId>
+            <artifactId>org.wso2.carbon.identity.application.mgt</artifactId>
```
can't add this dependency. It will create a cyclic dependency.

johannnallathamby commented on components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/listener/IdentityProviderMgtApplicationListener.java:
```diff
@@ -0,0 +1,123 @@
+                    .getApplicationDAO().getAllApplicationBasicInfo();
+            String tenantDomain = CarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+
+            for (int i = 0; i < applicationBasicInfos.length; i++) {
```
Use foreach

johannnallathamby commented on components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/listener/IdentityProviderMgtApplicationListener.java:
```diff
@@ -0,0 +1,123 @@
+                }
+            }
+        } catch (IdentityApplicationManagementException | IdentityException e) {
+            log.error("Error when updating default authenticator of service providers", e);
```
Don't log. Only throw.

johannnallathamby commented on components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/listener/IdentityProviderMgtProvisioningListener.java:
```diff
@@ -45,36 +46,61 @@
     @Override
-    public void updateResidentIdP(IdentityProvider identityProvider) {
+    public boolean doPreUpdateResidentIdP(IdentityProvider identityProvider) throws IdentityProviderManagementException {
         log.debug("update Resident Identity Provider event received");
```
Check isDebugEnabled()

johannnallathamby commented on components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/listener/IdentityProviderMgtProvisioningListener.java:
```diff
@@ -32,49 +32,77 @@
-        log.debug("update Resident Identity Provider event received");
+    public boolean doPreUpdateResidentIdP(IdentityProvider identityProvider) throws IdentityProviderManagementException {
+        if(log.isDebugEnabled()){
+            log.debug("update Resident Identity Provider event received");
```
Remove the unnecessary debug log

johannnallathamby commented:
Can we add a Abstract listener for both IdentityProviderMgtLister and ApplicationMgtListener and provide default implementation for all methods that do nothing, but just return true. That way we can avoid implementing the same default implementation over and over again in the concrete listeners.

johannnallathamby commented on components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/listener/IdentityProviderMgtProvisioningListener.java:
```diff
@@ -32,49 +32,77 @@
     @Override
-    public void addIdP(IdentityProvider identityProvider) {
+    public boolean doPreAddIdP(IdentityProvider identityProvider) throws IdentityProviderManagementException {
         log.debug("add new Identity Provider event received");
```
No isDebugEnabled()

johannnallathamby commented on components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/listener/IdentityProviderMgtProvisioningListener.java:
```diff
@@ -32,49 +32,77 @@
         } catch (IdentityProvisioningException e) {
-            log.error(e.getMessage(), e);
+            log.error("Error when provisioning IDP deletion", e);
+            throw new IdentityProviderManagementException("Error when provisioning IDP deletion",e);
```
Don't do log and throw. Just throw is enough.

johannnallathamby commented on components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/listener/IdentityProviderMgtProvisioningListener.java:
```diff
@@ -32,49 +32,77 @@
             destroyConnector(oldIdPName, tenantDomain);
         } catch (IdentityProvisioningException e) {
-            log.error(e.getMessage(), e);
+            log.error("Error when provisioning IDP update", e);
```
log and throw anti patttern

johannnallathamby commented on components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/listener/ApplicationMgtProvisioningListener.java:
```diff
@@ -33,23 +34,39 @@
-    public void deleteApplication(String applicationName) {
+    public boolean doPreDeleteApplication(String applicationName) throws IdentityApplicationManagementException {
         String tenantDomain = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
         log.debug("Clearing cache entry for " + applicationName);
```
No isDebugEnabled()

johannnallathamby commented on components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/listener/ApplicationMgtProvisioningListener.java:
```diff
@@ -33,23 +34,39 @@
-    public void updateApplication(ServiceProvider serviceProvider) {
+    public boolean doPreUpdateApplication(ServiceProvider serviceProvider) throws IdentityApplicationManagementException {
         String tenantDomain = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
         log.debug("Clearing cache entry for " + serviceProvider.getApplicationName());
```
No isDebugEnabled()

johannnallathamby commented on components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/listener/IdentityProviderMgtProvisioningListener.java:
```diff
@@ -32,49 +32,77 @@
             destroyConnector(idPName, tenantDomain);
         } catch (IdentityProvisioningException e) {
-            log.error(e.getMessage(), e);
+            log.error("Error when provisioning IDP deletion", e);
```
log and throw anti pattern. Only throw is enough

johannnallathamby commented on components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/IdentityProviderManagementService.java:
```diff
@@ -120,10 +126,10 @@ public void updateResidentIdP(IdentityProvider identityProvider)
     public IdentityProvider getIdPByName(String idPName)
-            throws IdentityApplicationManagementException {
+            throws IdentityProviderManagementException {
         if (StringUtils.isEmpty(idPName)) {
```
Can check isBlank()

johannnallathamby commented on components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/ConfigurationFacade.java:
```diff
@@ -90,7 +91,7 @@ public ExternalIdPConfig getIdPConfigByName(String idpName, String tenantDomain)
             }
-        } catch (IdentityApplicationManagementException e) {
+        } catch (IdentityProviderManagementException e) {
             log.error("Exception while getting IdP by name", e);
```
Ideally we must be throwing this exception and its the client who can decide what to do with it. Lets change this to throw the exception and in authentication framework we can catch the exception, log it and continue.

johannnallathamby commented on components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/ConfigurationFacade.java:
```diff
@@ -123,7 +124,7 @@ public ExternalIdPConfig getIdPConfigByRealm(String realm, String tenantDomain)
             }
-        } catch (IdentityApplicationManagementException e) {
+        } catch (IdentityProviderManagementException e) {
             log.error("Exception while getting IdP by realm", e);
```
Ideally we must be throwing this exception and its the client who can decide what to do with it. Lets change this to throw the exception and in authentication framework we can catch the exception, log it and continue.

johannnallathamby commented on components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java:
```diff
@@ -105,28 +105,30 @@ public static ApplicationManagementServiceImpl getInstance() {
-            for (ApplicationMgtListener listener : listeners) {
-                listener.createApplication(serviceProvider);
+        for (ApplicationMgtListener listener : listeners) {
+            if (!listener.doPreCreateApplication(serviceProvider)) {
```
We may need to change the API for the listener to take tenant domain and username (as arguments) similar to the createApplication method in ApplicationManagementServiceImpl. Ideally the listener interfaces should be same as the service interface. Currently the listener doesn't get the tenant domain and username. It is however set in CarbonContext because we are starting the tenant flow. However when it comes to the post listeners further down you can see we have already ended the tenant flow, and now there is no way to find the tenant domain or username. So basically the API for the listener is wrong, and we need to fix it.

johannnallathamby commented on components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java:
```diff
@@ -142,6 +144,12 @@ public int createApplication(ServiceProvider serviceProvider, String tenantDomai
         }
+
+        for (ApplicationMgtListener listener : listeners) {
+            if (!listener.doPostCreateApplication(serviceProvider)) {
```
At this point we have already ended the tenant flow and no way of knowing the tenant domain or username that was passed to the service method.
